src/extension.ts

```diff
@@ -57,44 +62,62 @@ export async function activate(ctx: vscode.ExtensionContext) {
+        await languageServerClient.start(ctx);
+    } else {
+        Command.dynamicRegister(languageServerClient.installLSPCommandName, LSPNotEnabledCommandHandler);
     }
+    Command.RegisteredCommands.push(languageServerClient.installLSPCommandName); // TODO: work out better way to do this..
+
+    ctx.subscriptions.push(new LintCommand(ctx))
+    if (getConfiguration().indexing.enabled) {
+            ctx.subscriptions.push(
+            new PlanCommand(runner, indexAdapter, ctx),
+            new IndexCommand(indexAdapter, ctx),
+            new ValidateCommand(indexAdapter, runner, ctx),
+            new ShowReferencesCommand(indexAdapter, ctx),
+            new NavigateToSectionCommand(indexAdapter, ctx),
+            new PreviewGraphCommand(indexAdapter, runner, ctx),
+            new ReindexCommand(indexAdapter, watcher, ctx));
+            // providers
+            vscode.languages.registerCompletionItemProvider(documentSelector, new CompletionProvider(indexAdapter), '.', '"', '{', '(', '['),
+            vscode.languages.registerDefinitionProvider(documentSelector, new DefinitionProvider(indexAdapter)),
+            vscode.languages.registerDocumentSymbolProvider(documentSelector, new DocumentSymbolProvider(indexAdapter)),
+            vscode.languages.registerWorkspaceSymbolProvider(new WorkspaceSymbolProvider(indexAdapter)),
+            vscode.languages.registerReferenceProvider(documentSelector, new ReferenceProvider(indexAdapter)),
+            vscode.languages.registerRenameProvider(documentSelector, new RenameProvider(indexAdapter)),
+            vscode.languages.registerHoverProvider(documentSelector, new HoverProvider(indexAdapter)),
+            vscode.languages.registerDocumentLinkProvider(documentSelector, new DocumentLinkProvider(indexAdapter)),
+            vscode.languages.registerFoldingRangeProvider(documentSelector, new CodeFoldingProvider(indexAdapter))
+            // views
+            vscode.window.registerTreeDataProvider('terraform-modules', new ModuleOverview(indexAdapter));
+            if (getConfiguration().codelens.enabled) {
+                ctx.subscriptions.push(vscode.languages.registerCodeLensProvider(documentSelector, new CodeLensProvider(indexAdapter)));
+            }
+            // operations which should only work in a local context (as opposed to live-share)
+            if (vscode.workspace.rootPath) {
+                ctx.subscriptions.push(vscode.workspace.onDidChangeTextDocument((e) => liveIndex(indexAdapter, e)));
+                // start to build the index
+                if (watcher) {
+                    await watcher.crawl();
+                }
+            }
+    } else {
+        const IndexerNotEnabledCommandHandler = () => {
+            vscode.window.showErrorMessage('Cannot perform action currently using the Terraform Language Server not Indexer.');
+        };
+        Command.dynamicRegister(PlanCommand.CommandName, IndexerNotEnabledCommandHandler);
```

Comment:
It makes sense to disable these commands (for now) and bring them back later. Currently the only reason why the `plan` command needs the `IndexAdapter` is to figure out which terraform version is required by the group (`PlanCommand.getExecutable` calls `Runner.getRequiredExecutable`).

I wonder how this can easiest be solved going forward? (@juliosueiras might wanna chime in) It would be nice to be able to know something about the templates from within the vscode-terraform plugin, I am unsure where to draw the line though.

It would be nice if we don't introduce too many dependencies between the plugin and the language-server as that makes deployment and releasing more complicated (in order to simplify things in the future we should introduce a version-check).

Also, I do not know a lot about the language-server-protocol, but I guess we can introduce our own calls so that we can communicate with the server outside of what vscode does itself?

Reply:
@mauve just wanted to make sure for this part " I guess we can introduce our own calls so that we can communicate with the server outside of what vscode does itself?", can you give a example of that call? since I am unsure what you mean by that

Reply:
@mauve the LSP has a well defined interface, outside of that it's up to the extension to do the work. https://github.com/Microsoft/language-server-protocol

I'll add an issue to investigate re-instating `plan`, `validate` and other commands that can be picked up as a bit of followup work, worried that if I take on too much in this PR it won't get finished. Sound like a good plan?

Reply:
@juliosuerias i was thinking how to implement the plan command. I guess there are two options:

1. Add the command directly to the LSP (if possible)

2. Implement the plan command in the plugin and Expose a custom interaction between the plugin and the LSP which tells us which terraform version is required by a set of templates

This feature is a bit of a corner case, but I implemented it because I have different versions of terraform installed because of conflicting requirements for different templates but this allowed me to have them in the same vscode workspace.

Reply:
@mauve for the lsp side, I can look into implementing either [Execute Command](https://microsoft.github.io/language-server-protocol/specification#workspace_executeCommand) or [Code action](https://microsoft.github.io/language-server-protocol/specification#textDocument_codeAction),
I am also thinking of maybe allowing plan on selected resource/data source? what you think?

Reply:
@mauve I've found tfenv to be the best tool for handling multiple versions of Terraform

Reply:
@smiller171 `tfenv` thanks for the tip, it looks fine but has a bash dependency and therefore only works on windows using git-bash or wsl (that feels like a show-stopper for us). (also it seems it cheats when picking the version).

Reply:
@juliosueiras aha so that means you can implement the plan command directly (as is) in the LSP. I don't know what the best way forward is to be honest moving as much as possible into the LSP or having a split.

The benefit of writing features directly in the plugin is that it is relatively easy to debug and develop and that I can stacktraces and exceptions reported to my Azure AppInsights.

The benefit of writing features in the LSP is that is where all the data is.

It might be possible to do a split but quite honestly it seems like having everything in one place is probably better.

For now though I think we should wait with `plan` and friends and focus on doing one release with the LSP to get data. If we can initialize the LSP with the AppInsights key you could use the AppInsights-Go SDK to track exceptions in the LSP. I can pay for the metrics and give you guys access to my dashboard with stats and crash data (beware that I have currently disabled the AppInsights stuff but can enable again when we do the first release with the LSP).